llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,145 @@
+  auto *Literal = cast<Constant>(CI->getArgOperand(0));
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return CI->getModule()->getOrInsertGlobal(
+      CASLiteralName, CASLiteralType, [&] {
+        StringRef LiteralValue;
+        getConstantStringInfo(Literal, LiteralValue);
+        GlobalVariable *GV = Builder.CreateGlobalString(
+            LiteralValue, CASLiteralName, ConstantAddrspaceID, CI->getModule());
+        GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+        GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+        return GV;
+      });
+}
+
+FunctionCallee getCASPrintfFunction(Function &GenericASPrintfFunc,
+                                    Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      GenericASPrintfFunc.getParent()->getOrInsertFunction(
+          "_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+          GenericASPrintfFunc.getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+CallInst *buildCASPrintfCall(IRBuilder<> &Builder, FunctionCallee CASPrintfFunc,
+                             Constant *CASLiteral, CallInst *CI) {
+  SmallVector<Value *, 4> CallOperands(CI->data_operands_begin(),
+                                       CI->data_operands_end());
+  CallOperands[0] = CASLiteral;
+  Builder.SetInsertPoint(CI);
+  CallInst *CASCall = Builder.CreateCall(CASPrintfFunc, CallOperands,
+                                         CI->getName().str() + "._AS2");
+  CASCall->setTailCall(true);
+  return CASCall;
+}
+} // namespace
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  size_t ReplacedCallsCount = 0;
+  SmallVector<Function *, 1> FunctionsToDrop;
+
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+    if (!F.getName().contains("__spirv_ocl_printf"))
+      continue;
+    auto *LiteralType = F.getArg(0)->getType();
+    if (LiteralType->getPointerAddressSpace() == ConstantAddrspaceID)
+      // No need to replace the literal type and its printf users
+      continue;
+
+    IRBuilder<> Builder(M.getContext());
+    SmallVector<CallInst *, 8> CallInstsToDrop;
+    for (User *U : F.users()) {
+      if (!isa<CallInst>(U))
+        continue;
+      auto *CI = cast<CallInst>(U);
+      Constant *CASLiteral = getCASLiteral(Builder, CI);
+      FunctionCallee CASPrintfFunc =
+          getCASPrintfFunction(F, CASLiteral->getType());
+      CI->replaceAllUsesWith(
```

Comment:
I've suggested to use this API myself, but now I wonder how does it work if there are no uses? Most of `printf` users don't care about its return value.

Reply:
As far as I can tell from local testing, there's no problem with calling the API for the non-used instruction - it's simply that "all of the zero uses" get replaced. From a brief glance, the API's implementation code doesn't contradict this notion: https://llvm.org/doxygen/Value_8cpp_source.html#l00501

Reply:
> As far as I can tell from local testing, there's no problem with calling the API for the non-used instruction - it's simply that "all of the zero uses" get replaced. From a brief glance, the API's implementation code doesn't contradict this notion: https://llvm.org/doxygen/Value_8cpp_source.html#l00501

Nevermind, I've just realized that we explicitly insert a new call instruction into a module, so it will always be there even if there are no uses of it